src/scope.rs

```diff
@@ -83,6 +84,62 @@ impl Scope {
             factory_ref: fref,
         }
     }
+
+
+    /// Run external configuration as part of the scope building
+    /// process
+    ///
+    /// This function is useful for moving parts of configuration to a
+    /// different module or even library. For example,
+    /// some of the resource's configuration could be moved to different module.
+    ///
+    /// ```rust
+    /// # extern crate actix_web;
+    /// use actix_web::{web, middleware, App, HttpResponse};
+    ///
+    /// // this function could be located in different module
+    /// fn config(cfg: &mut web::ScopeConfig) {
+    ///     cfg.service(web::resource("/test")
+    ///         .route(web::get().to(|| HttpResponse::Ok()))
+    ///         .route(web::head().to(|| HttpResponse::MethodNotAllowed()))
+    ///     );
+    /// }
+    ///
+    /// fn main() {
+    ///     let app = App::new()
+    ///         .wrap(middleware::Logger::default())
+    ///         .service(
+    ///             web::scope("/api")
+    ///                 .configure(config)
+    ///         )
+    ///         .route("/index.html", web::get().to(|| HttpResponse::Ok()));
+    /// }
+    /// ```
+    pub fn configure<F>(mut self, f: F) -> Self
+        where
+            F: FnOnce(&mut ScopeConfig),
+    {
+        let mut cfg = ScopeConfig::new();
+        f(&mut cfg);
+        self.services.extend(cfg.services);
+        self.guards.extend(cfg.guards);
+        //self.data.extend(cfg.data);
+
+        if cfg.data.is_some() {
```

Comment:
What is the reason?

Reply:
Because of this:

https://github.com/actix/actix-web/blob/ddce1725725bfc89a1e385b1a366488b1c2d79e5/src/scope.rs#L63-L66

`data` is an `Option<Extensions>` in Scope, but a `Vec<Box<DataFactory>>` in App, therefore the configure function must handle the case where the `Option` is `None`, and merge accordingly.

Am I wrong?

https://github.com/actix/actix-web/blob/ddce1725725bfc89a1e385b1a366488b1c2d79e5/src/app.rs#L34

Reply:
For reference:
https://github.com/actix/actix-web/blob/ddce1725725bfc89a1e385b1a366488b1c2d79e5/src/app.rs#L29-L38

vs:
https://github.com/actix/actix-web/blob/ddce1725725bfc89a1e385b1a366488b1c2d79e5/src/scope.rs#L63-L71

Reply:
Whilst we can do this in `App`:
https://github.com/actix/actix-web/blob/ddce1725725bfc89a1e385b1a366488b1c2d79e5/src/app.rs#L129-L139

We can't do it in `Scope` because of the `Option` type:
https://github.com/actix/actix-web/blob/ddce1725725bfc89a1e385b1a366488b1c2d79e5/src/scope.rs#L118-L143